### PR TITLE
Disallow changes for content moderation state for locked pages

### PR DESCRIPTION
**Pull request by balagan73**
#644 

- Checking if content_moderation module is enabled, then resetting published state, if it was the original.

### CODE REVIEW

**Comment by balagan73** [comment]
We don't have content moderation in DS, but as we sync this code from here to projects, where it might be installed, I thought it is worth to add it here.

**Review by andreytroeglazov — APPROVED** [review]

File: web/modules/custom/server_general/server_general.module [hunk 1]
```diff
@@ -154,8 +154,17 @@ function server_general_node_presave(NodeInterface $entity) {
     return;
   }
 
+  // Get the module handler service.
+  $module_handler = \Drupal::service('module_handler');
   /** @var \Drupal\node\NodeInterface $original */
   $original = $entity->original;
+  // If content moderation is used, don't allow status change for locked pages.
+  if ($module_handler->moduleExists('content_moderation')) {
+    if ($original->get('moderation_state')->getString() === 'published' && $entity->get('moderation_state')->getString() !== 'published') {
```
Comment (amitaibu) [inline web/modules/custom/server_general/server_general.module:163]: Can be in a single IF

File: web/modules/custom/server_general/server_general.module [hunk 2]
```diff
@@ -154,8 +154,15 @@ function server_general_node_presave(NodeInterface $entity) {
     return;
   }
 
+  // Get the module handler service.
+  $module_handler = \Drupal::service('module_handler');
   /** @var \Drupal\node\NodeInterface $original */
   $original = $entity->original;
+  // If content moderation is used, don't allow unpublish for locked pages.
+  if ($module_handler->moduleExists('content_moderation') && $original->get('moderation_state')->getString() === 'published' && $entity->get('moderation_state')->getString() === 'unpublished') {
```
Comment (amitaibu) [inline web/modules/custom/server_general/server_general.module:162]: Maybe better?

```suggestion
  if ($module_handler->moduleExists('content_moderation') && $original->get('moderation_state')->getString() === 'published' && $entity->get('moderation_state')->getString() !== 'published') {
```

Comment (balagan73) [reply web/modules/custom/server_general/server_general.module:162]: In another project the specific request was to only disallow unpublishing, as draft etc creates new revisions. I think it makes sense.

**Comment by balagan73** [comment]
I have created a module with the name server_locked_pages, to help syncing changes related to locked pages.

Delete option is not displayed|Delete path cannot be accessed
-|-
image|image

**Comment by amitaibu** [comment]
> I have created a module with the name server_locked_pages

Let's keep it inside server_general. Later, we'll move it out into a contrib module

File: web/modules/custom/server_general/server_general.module (hunk 2 above)
Comment (balagan73) [reply web/modules/custom/server_general/server_general.module:162]: @amitaibu This is the only question now, whether we should disallow only unpublish or the other states too. As I noted above, I think disallowing only unpublish is fine.

Comment (amitaibu) [reply web/modules/custom/server_general/server_general.module:162]: >  As I noted above, I think disallowing only unpublish is fine.

:+1:

**Pull request merged**